azurelinuxagent/ga/agent_update.py

```diff
@@ -0,0 +1,340 @@
+                    if os.path.isfile(agent_path):
+                        logger.info(u"Purging outdated Agent file {0}", agent_path)
+                        os.remove(agent_path)
+                    else:
+                        logger.info(u"Purging outdated Agent directory {0}", agent_path)
+                        shutil.rmtree(agent_path)
+            except Exception as e:
+                logger.warn(u"Purging {0} raised exception: {1}", agent_path, ustr(e))
+
+    @staticmethod
+    def __proceed_with_update(requested_version):
+        """
+        If requested version is specified, upgrade/downgrade to the specified version.
+        Raises: AgentUpgradeExitException
+        """
+        if requested_version < CURRENT_VERSION:
+            # In case of a downgrade, we mark the current agent as bad version to avoid starting it back up ever again
+            # (the expectation here being that if we get request to a downgrade,
+            # there's a good reason for not wanting the current version).
+            prefix = "downgrade"
+            try:
+                # We should always have an agent directory for the CURRENT_VERSION
+                agents_on_disk = AgentUpdateHandler.__get_available_agents_on_disk()
+                current_agent = next(agent for agent in agents_on_disk if agent.version == CURRENT_VERSION)
+                msg = "Marking the agent {0} as bad version since a downgrade was requested in the GoalState, " \
+                      "suggesting that we really don't want to execute any extensions using this version".format(CURRENT_VERSION)
+                logger.info(msg)
+                current_agent.mark_failure(is_fatal=True, reason=msg)
+            except StopIteration:
+                logger.warn(
+                    "Could not find a matching agent with current version {0} to blacklist, skipping it".format(
+                        CURRENT_VERSION))
+        else:
+            # In case of an upgrade, we don't need to exclude anything as the daemon will automatically
+            # start the next available highest version which would be the target version
+            prefix = "upgrade"
+        raise AgentUpgradeExitException("Agent update found, Exiting current process to {0} to the new Agent version {1}".format(prefix, requested_version))
+
+    @staticmethod
+    def __get_available_agents_on_disk():
+        available_agents = [agent for agent in AgentUpdateHandler.__get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    @staticmethod
+    def __get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if os.path.isdir(agent_dir)]
+
+    def __log_event(self, level, msg_, success_=True):
```

Comment:
Why the trailing underscore in msg_ and success_? the convention in Python is to do that when using a name that collides with a built-in name, but that does not seem to be the case here